---
title: "Some karyoploteR use cases"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Some karyoploteR use cases}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width=11,
  fig.height=10,
  results = 'hide',
  message=F
)
```

```{r setup}
library(pidProjCode)
library(karyoploteR)
library(data.table)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
```

As sample data sets we use subsets of four GWAS: one of selective IgA deficiency by Bronson et al. (2016),  another of primary immunodeficiency by Thaventhiran et al. (2020), a third of systemic lupus erythematosus by Bentham et al. (2015), and a fourth of ulcerative colitis by de Lange et al. (2017). These contain only SNPs on chromosome 6 from 25Mb to 36Mb (this interval includes the MHC locus).

```{r loadExampleData}
igadDat <- fread(file = system.file('extdata', 'igadGwas.tsv.gz', package = 'pidProjCode', mustWork = T), sep = '\t', header = T, select = c('SNPID','CHR38','BP38','P'))
pidDat <- fread(file = system.file('extdata', 'pidGwas.tsv.gz', package = 'pidProjCode', mustWork = T), sep = '\t', header = T, select = c('SNPID','CHR38','BP38','P'))
sleDat <- fread(file = system.file('extdata', 'sleGwas.tsv.gz', package = 'pidProjCode', mustWork = T), sep = '\t', header = T, select = c('SNPID','CHR38','BP38','P'))
ucDat <- fread(file = system.file('extdata', 'ucGwas.tsv.gz', package = 'pidProjCode', mustWork = T), sep = '\t', header = T, select = c('SNPID','CHR38','BP38','P'))

# ymax doesn't seem to work for me with karyoploteR
igadDat <- igadDat[-log10(P) < 15]
pidDat <- pidDat[-log10(P) < 15]
sleDat <- sleDat[-log10(P) < 15]
ucDat <- ucDat[-log10(P) < 15]

# GenomicRanges require that chromosome seqnames have the prefix 'chr'
igadDat[ , CHR38 := paste0('chr', CHR38)]
pidDat[ , CHR38 := paste0('chr', CHR38)]
sleDat[ , CHR38 := paste0('chr', CHR38)]
ucDat[ , CHR38 := paste0('chr', CHR38)]

```

`karyoploteR` requires you pass genomic data in the form `GenomicRanges` objects. The `GenomicRanges` package provides the `makeGRangesFromDataFrame` function to do this. (Note that the whole point of this vignette was to explain how to use my own `createGRanges` function but that has since been obviated by my discovery that use of `makeGRangesFromDataFrame` is more straightforward than originally thought.)

A useful feature of `karyoploteR` is the highlighting of specified SNPs. These need to be given in a `GenomicRanges` object, too, so we create `topSnpGRanges` to hold these (in this case just a single SNP). For a Manhattan plot we want a y axis on the negative log10 scale, but `karyoploteR::kpPoints` doesn't seem to heed this rescaling, so we specify the y coordinate as an extra metadata column in the `GenomicRanges` object

```{r creatingGRangesObjects}
igadGRanges <- makeGRangesFromDataFrame(data.frame(igadDat), start.field = 'BP38', end.field = 'BP38', seqnames.field = 'CHR38', ignore.strand = T, keep.extra.columns = T)
pidGRanges <- makeGRangesFromDataFrame(data.frame(pidDat), start.field = 'BP38', end.field = 'BP38', seqnames.field = 'CHR38', ignore.strand = T, keep.extra.columns = T)
sleGRanges <- makeGRangesFromDataFrame(data.frame(sleDat), start.field = 'BP38', end.field = 'BP38', seqnames.field = 'CHR38', ignore.strand = T, keep.extra.columns = T)
ucGRanges <- makeGRangesFromDataFrame(data.frame(ucDat), start.field = 'BP38', end.field = 'BP38', seqnames.field = 'CHR38', ignore.strand = T, keep.extra.columns = T)

topSnps <- c('rs116540075')

topSnpGRanges <- makeGRangesFromDataFrame(data.frame(igadDat[SNPID %in% topSnps][, y:= -log10(P)]), start.field = 'BP38', end.field = 'BP38', seqnames.field = 'CHR38', ignore.strand = T, keep.extra.columns = T)

```

The first reference for use of `karyoploteR` should be the docs themselves. In particular, I relied on the author's [Manhattan tutorial](https://bernatgel.github.io/karyoploter_tutorial//Tutorial/PlotManhattan/PlotManhattan.html) but some of the details are now outdated, such as its use of the annotations from the `hg19` draft genome. Look at that page first to see the full array of Manhattan plot functionality offered by this package.

I spent some hours muddling through the process of adapting the examples given in the tutorial to my own use case. I hope this short vignette serves to motivate the use of the `createGRanges` function ([the source code for which you can just copy from my package's GitHub repo](https://github.com/twillis209/pidProjCode/blob/master/R/createGRanges.R)) and to provide two examples to copy and adapt for the 'zoomed-in Manhattan plot' use case. Copying my code rather than that in the tutorial should hopefully spare you some of the trouble I went through getting the latter to work.


```{r manhattan}
ymax <- 137
r0 <- 0
cex.main <- 1.3
cex.tick <- 0.75

kp <- plotKaryotype(plot.type=4, labels.plotter=NULL, zoom="chr6:25e6-36e6")
title(main='IgAD GWAS, chr6:25M-36M, MHC boxed in blue, top extra-MHC SNP rs116540075 circled in red', cex.main=cex.main)
kpAddBaseNumbers(kp, add.units=T, cex=cex.tick, tick.dist=5e5)
kpAxis(kp, ymin=0,ymax=ymax, r0=r0)

# karyoploteR looks for a column with the name 'p' or 'P' in the metadata columns of the GenomicRanges object. Alternatively, y coordinates can be passed as a vector to the argument 'pval' (see the docs).
kp <- kpPlotManhattan(kp, data=igadGRanges, points.col='brewer.set3', ymax=ymax, r0=r0)
kpRect(kp, chr='chr6', x0=28510120, x1=33480577,r0=r0, y0=0,y1=1, col=NA, border='blue', lwd=3)
kp <- kpPoints(kp, data=topSnpGRanges, pch=1, cex=1.6, col="red", lwd=2, ymax=ymax, r0=r0)
```

```{r zoomedManhattan}
ymax <- 38
r0 <- 0.2
cex.tick <- 1

kp <- plotKaryotype(plot.type=4, labels.plotter=NULL, zoom="chr6:33e6-34e6")
title(main='IgAD GWAS, chr6:33M-34M, neighbourhood of top extra-MHC SNP rs116540075 (circled in red)', cex.main=cex.main)
kpAddBaseNumbers(kp, add.units=T, cex=cex.tick, tick.dist=1e5)
kpAxis(kp, ymin=0,ymax=ymax, r0=r0)
kp <- kpPlotManhattan(kp, data=igadGRanges, points.col='brewer.set3', ymax=ymax, r0=r0)
kpRect(kp, chr='chr6', x0=33e6, x1=33480577,r0=r0, y0=0,y1=1, col=NA, border='blue', lwd=3)
kp <- kpPoints(kp, data=topSnpGRanges, pch=1, cex=1.6, col="red", lwd=2, ymax=ymax, r0=r0)

genes.data <- makeGenesDataFromTxDb(karyoplot=kp, txdb=TxDb.Hsapiens.UCSC.hg38.knownGene)

genes.data <- addGeneNames(genes.data)
genes.data.merged <- mergeTranscripts(genes.data)
kp <- kpPlotGenes(kp, data=genes.data.merged, r0=0, r1=0.2, cex=1.0, gene.name.position='left')
```

```{r backToBackManhattans}
cex.tick <- 1
cex.label <- 1.8

kp <- plotKaryotype(plot.type=4, labels.plotter=NULL, zoom="chr6:25e6-36e6")
kpAddBaseNumbers(kp, add.units=T, cex=cex.tick, tick.dist=1e6)


title(main='IgAD and PID GWAS, chr6:25M-36M', cex.main=cex.main)

kpAddLabels(kp, labels = "IgAD", srt=90, pos=3, r0=0.7, r1=1, cex=cex.label, label.margin = 0.025)
kpAxis(kp, ymin=0, ymax=40, r0=0.5)
kp <- kpPlotManhattan(kp, data=igadGRanges, r0=0.5, r1=1, ymax=40)
kpAddLabels(kp, labels = "PID", srt=90, pos=3, r0=0, r1=0.3, cex=cex.label, label.margin = 0.025)
kpAxis(kp, ymin=0, ymax=10, r0=0.5, r1=0)
kp <- kpPlotManhattan(kp, data=pidGRanges, r0=0.5, r1=0, ymax=10, points.col = "2blues")
```

```{r stackedManhattans}
pidGRanges<-subset(pidGRanges, -log10(P) < 10)
sleGRanges<-subset(sleGRanges, -log10(P) < 10)
ucGRanges<-subset(ucGRanges, -log10(P) < 10)
igadGRanges<-subset(igadGRanges, -log10(P) < 10)

cex.tick <- 1
cex.label <- 1
ymax <- 10
label.margin <- 0.03
# The % of the total plotting space to leave in each margin between the tracks
autotrack.margin <- 0.1

kp <- plotKaryotype(plot.type=4, labels.plotter=NULL, zoom="chr6:25e6-36e6")
kpAddBaseNumbers(kp, add.units=T, cex=cex.tick, tick.dist=1e6)

# If plotting genome-wide loci, replace the call to kpAddBaseNumbers with the line below
# kpAddChromosomeNames(kp, col='black', srt=90, cex=2)

title(main='IgAD, PID, UC, and SLE GWAS, chr6:25M-36M', cex.main=cex.main)

# Note that the last keyword argument, which specifies that this plot should occupy position 1 of 4
# kpPlotManhattan will look for a column labelled 'p' or 'P' in the GenomicRanges object, but you can specify it explicitly by passing it as a vector to the 'pval' argument

# Note the points.col argument; I recommend brewer.set3 when looking at the whole genome because the points from different chromosomes will be coloured distinctly, although we can't see that here as I have data from only one chromosome

at<-autotrack(1, 4, margin=autotrack.margin)
kp <- kpPlotManhattan(kp, data=pidGRanges, pval=pidGRanges$P, points.col = 'brewer.set3', r0=at$r0, r1=at$r1, ymax=ymax)
kpAddLabels(kp, labels = 'PID', srt=90, pos=3, r0=at$r0, r1=at$r1, cex=cex.label, label.margin=label.margin)
kpAxis(kp, ymin=0, ymax=ymax, r0=at$r0, r1=at$r1)

at<-autotrack(2, 4, margin=autotrack.margin)
kp <- kpPlotManhattan(kp, data=igadGRanges, points.col = 'brewer.set3', r0=at$r0, r1=at$r1, ymax=ymax)
kpAddLabels(kp, labels = 'IgAD', srt=90, pos=3, r0=at$r0, r1=at$r1, cex=cex.label, label.margin=label.margin)
kpAxis(kp, ymin=0, ymax=ymax, r0=at$r0, r1=at$r1)

at<-autotrack(3, 4, margin=autotrack.margin)
kp <- kpPlotManhattan(kp, data=sleGRanges, points.col = 'brewer.set3', r0=at$r0, r1=at$r1, ymax=ymax)
kpAddLabels(kp, labels = 'SLE', srt=90, pos=3, r0=at$r0, r1=at$r1, cex=cex.label, label.margin=label.margin)
kpAxis(kp, ymin=0, ymax=ymax, r0=at$r0, r1=at$r1)

at<-autotrack(4, 4, margin=autotrack.margin)
kp <- kpPlotManhattan(kp, data=ucGRanges, points.col = 'brewer.set3', r0=at$r0, r1=at$r1, ymax=ymax)
kpAddLabels(kp, labels = 'UC', srt=90, pos=3, r0=at$r0, r1=at$r1, cex=cex.label, label.margin=label.margin)
kpAxis(kp, ymin=0, ymax=ymax, r0=at$r0, r1=at$r1)
```

```{r stackedManhattansWhole}
kp <- plotKaryotype(plot.type=4, labels.plotter=NULL)
kpAddChromosomeNames(kp, col='black', srt=90, cex=1)

title(main='IgAD, PID, UC, and SLE GWAS', cex.main=cex.main)

at<-autotrack(1, 4, margin=autotrack.margin)
kp <- kpPlotManhattan(kp, data=pidGRanges, pval=pidGRanges$P, points.col = 'brewer.set3', r0=at$r0, r1=at$r1, ymax=ymax)
kpAddLabels(kp, labels = 'PID', srt=90, pos=3, r0=at$r0, r1=at$r1, cex=cex.label, label.margin=label.margin)
kpAxis(kp, ymin=0, ymax=ymax, r0=at$r0, r1=at$r1)

at<-autotrack(2, 4, margin=autotrack.margin)
kp <- kpPlotManhattan(kp, data=igadGRanges, points.col = 'brewer.set3', r0=at$r0, r1=at$r1, ymax=ymax)
kpAddLabels(kp, labels = 'IgAD', srt=90, pos=3, r0=at$r0, r1=at$r1, cex=cex.label, label.margin=label.margin)
kpAxis(kp, ymin=0, ymax=ymax, r0=at$r0, r1=at$r1)

at<-autotrack(3, 4, margin=autotrack.margin)
kp <- kpPlotManhattan(kp, data=sleGRanges, points.col = 'brewer.set3', r0=at$r0, r1=at$r1, ymax=ymax)
kpAddLabels(kp, labels = 'SLE', srt=90, pos=3, r0=at$r0, r1=at$r1, cex=cex.label, label.margin=label.margin)
kpAxis(kp, ymin=0, ymax=ymax, r0=at$r0, r1=at$r1)

at<-autotrack(4, 4, margin=autotrack.margin)
kp <- kpPlotManhattan(kp, data=ucGRanges, points.col = 'brewer.set3', r0=at$r0, r1=at$r1, ymax=ymax)
kpAddLabels(kp, labels = 'UC', srt=90, pos=3, r0=at$r0, r1=at$r1, cex=cex.label, label.margin=label.margin)
kpAxis(kp, ymin=0, ymax=ymax, r0=at$r0, r1=at$r1)
```

## Adding legends

As discussed [here](https://support.bioconductor.org/p/124328/), legends can be added as usual when working with base graphics.

```{r legendExample}

kp <- plotKaryotype(plot.type=4, labels.plotter=NULL, zoom="chr6:25e6-36e6")
kpAddBaseNumbers(kp, add.units=T, cex=cex.tick, tick.dist=1e6)


title(main='IgAD and PID GWAS, chr6:25M-36M', cex.main=cex.main)

kpAddLabels(kp, labels = "IgAD", srt=90, pos=3, r0=0.7, r1=1, cex=cex.label, label.margin = 0.025)
kpAxis(kp, ymin=0, ymax=40, r0=0.5)
kp <- kpPlotManhattan(kp, data=igadGRanges, r0=0.5, r1=1, ymax=40)
kpAddLabels(kp, labels = "PID", srt=90, pos=3, r0=0, r1=0.3, cex=cex.label, label.margin = 0.025)
kpAxis(kp, ymin=0, ymax=10, r0=0.5, r1=0)
kp <- kpPlotManhattan(kp, data=pidGRanges, r0=0.5, r1=0, ymax=10, points.col = "2blues")

legend(x = "topright", fill = c("#6caeff", "#2b5d9b"), legend = c("IgAD", "PID"))
```

## Using `backToBackManhattan`

I have written a convenience function for drawing 'back-to-back' Manhattan plots with `karyoploteR`, `backToBackManhattan`. These plots allow one to compare results from two GWAS. The function itself does nothing clever and was written merely to save my having to copy and tweak the same boilerplate plotting code over and over.

In my own work I have had cause to plot (transformed) GWAS results in this manner in order to assess two competing conditional FDR methods, `cfdr` and `fcfdr`. These condition p-values for the association of SNPs with a principal trait upon p-values for the association of the same SNPs with some informative auxiliary trait, and produce p-value-like 'v-values'. In this use case, it can be useful to have to hand the original principal trait p-values, so `backToBackManhattan` allows one to specify a third Manhattan.

With `backToBackManhattan` one can draw:
- whole-genome Manhattans
- results from a selection of chromosomes (see the `chromosomes` argument)
- a Manhattan which magnifies an interval on a single chromosome (see the `zoom` argument)

In addition, the `plotGenes` flag has `karyoploteR` draw gene structures on the lowermost track. At the moment this works only when the `zoom` argument is specified because it takes a long time to draw the genes and the larger the interval, the longer the plotting time. Varying the size of the interval in on which you have `zoom`ed should give you a sense of what is a sensible interval with respect to plotting time. Longer intervals also lead to the gene track getting rather busy, which is no help if you are trying to pick out the genes in which your SNPs of interest lie.

```{r wholeGenomeManhattan}
backToBackManhattan(pidGRanges, igadGRanges, topLabel='PID', bottomLabel='IgAD', main='PID and IgAD', axisLabelCex=1, axisLabelMargin=0.04)
```

```{r singleChromosomeManhattan}
backToBackManhattan(pidGRanges, igadGRanges, topLabel='PID', bottomLabel='IgAD', main='Chromosome 6', chromosomes='chr6', tickDist=1e7, axisLabelCex=1, axisLabelMargin=0.04)
```

```{r singleChromosomeThreeTracksManhattan}
backToBackManhattan(pidGRanges, igadGRanges, topLabel='PID', bottomLabel='IgAD', thirdLabel = 'SLE', thirdGRanges = sleGRanges, main='Chromosome 6', chromosomes='chr6', tickDist=1e7, axisLabelCex=1, axisLabelMargin=0.04)
```

```{r zoomedManhattanBackToBack}
backToBackManhattan(pidGRanges, igadGRanges, topLabel='PID', bottomLabel='IgAD', main='Chromosome 6, 25M-26M', zoom='chr6:25e6-36e6', tickDist=1e6, axisLabelCex=1, axisLabelMargin=0.04)
```

```{r zoomedManhattanWithGenes}
backToBackManhattan(pidGRanges, igadGRanges, topLabel='PID', bottomLabel='IgAD', main='Chromosome 6, 25M-26M', zoom='chr6:25e6-26e6', tickDist=1e5, plotGenes=T, axisLabelCex=1, axisLabelMargin=0.04)
```

```{r zoomedManhattanWithThreeTracksAndGenes}
backToBackManhattan(pidGRanges, igadGRanges, topLabel='PID', bottomLabel='IgAD', thirdGRanges=sleGRanges, thirdLabel='SLE', main='Chromosome 6, 25M-26M', zoom='chr6:25e6-26e6', tickDist=1e5, plotGenes=T, axisLabelCex=1, axisLabelMargin=0.04)
```